File: 92-Reverse-Linked-List-II.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseBetween(head *ListNode, left int, right int) *ListNode {
    var begin *ListNode
    for i := 1; i < left; i++ {
        if i == 1 {
            begin = head
        } else {
            begin = begin.Next
        }
    }
    front := head
    if begin != nil {
        front = begin.Next
    }
    prev := front
    node := prev.Next
    for i := left + 1; i <= right; i++ {
        next := node.Next
        node.Next = prev
        prev = node
        node = next
    }
    front.Next = node
    if begin != nil {
        begin.Next = prev
    } else {
        head = prev
    }
    return head
}

File: leetcode/92-Reverse-Linked-List-II.cpp
// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

ListNode *reverseBetween(ListNode *head, int left, int right) {
    // node before front
    ListNode *begin = nullptr;
    for (int i = 1; i < left; i++) {
        if (i == 1)
            begin = head;
        else
            begin = begin->next;
    }
    ListNode *front = begin ? begin->next : head;
    ListNode *prev = front;
    ListNode *node = prev->next;
    for (int i = left + 1; i <= right; i++) {
        ListNode *next = node->next;
        node->next = prev;
        prev = node;
        node = next;
    }
    // after loop, node == end, prev == back
    front->next = node;
    if (begin) {
        begin->next = prev;
    } else {
        head = prev;
    }
    return head;
}
